### PR TITLE
RFC: Add luks rootfs support

### PR DESCRIPTION
**Pull request by nandra**
Details are described here: https://hub.mender.io/t/rfc-mender-convert-luks-support/4158?u=marekbelisko

Thanks for comments/ideas.

### CODE REVIEW

File: modules/disk.sh [hunk 1]
```diff
@@ -108,11 +108,19 @@ disk_write_at_offset() {
...
+# $5 - partition need luks encryption
 disk_create_file_system_from_folder() {
-    log_info "Creating a file-system image from: ${1}"
+    log_info "Creating a file-system image from: ${1}, LUKS: ${5}"
 
     run_and_log_cmd "dd if=/dev/zero of=${2} seek=${3} count=0 bs=512 status=none"
 
+    if [ "${5}" = "y" ]; then
+        log_info "Creating LUKS encrypted partition"
+	echo -n "${LUKS_PWD}" | sudo cryptsetup -y -v luksFormat --pbkdf=pbkdf2 ${2} -d -
```
Comment (kacf) [inline modules/disk.sh:119]: `LUKS_PWD` comes from environment?

File: mender-convert-modify [hunk 2]
```diff
@@ -35,6 +35,10 @@ function trap_exit() {
     echo "mender-convert-modify has finished. Cleaning up..."
     sudo umount -f work/boot
     sudo umount -f work/rootfs
+    if [ ! -z ${loopdev} ]; then
```
Comment (vpodzime) [inline mender-convert-modify:38]: `! -z` is the same as `-n`

File: mender-convert-modify [hunk 3]
```diff
@@ -95,7 +99,18 @@ mkdir -p work/boot
...
-sudo mount ${root_part} work/rootfs
+
+luks=$(is_luks_partition ${root_part})
+log_info "LUKS: ${luks}"
+
+# handling for luks formatted partition
+if [ "${luks}" = "y" ]; then
+    # get loop device which will be used later to properly close/unmount/delete
+    loopdev=$(process_luks_partition "${root_part}")
+    sudo mount /dev/mapper/root work/rootfs
```
[the remark this answers is not in the document]
Comment (vpodzime) [reply mender-convert-modify:110]: There's `cryptsetup luksUUID` and it's a common practice to call the device mapper device as `luks-LUKS_UUID`.

File: mender-convert-package [hunk 4]
```diff
@@ -189,24 +205,25 @@ fi
...
 log_info "Rootfs filesystem size will be $(disk_sectors_to_mb ${rootfs_image_sectors}) MiB"
 
-# Extract file-system type from rootfs
-if file ${root_part} | grep -q ext4; then
+
+# Extract file-system type from mounted rootfs
+if sudo mount | grep "work/rootfs" | grep ext4; then
     image_fs_type="ext4"
-elif file ${root_part} | grep -q XFS; then
+elif sudo mount | grep -q "work/rootfs" | grep -q xfs; then
```
Comment (vpodzime) [inline mender-convert-package:212]: `mount` doesn't require `sudo`.

File: modules/disk.sh [hunk 5]
```diff
@@ -125,7 +133,11 @@ disk_create_file_system_from_folder() {
             if [ ! -f ${MKFS_EXT4} ]; then
                 MKFS_EXT4="/sbin/mkfs.ext4"
             fi
-            run_and_log_cmd "${MKFS_EXT4} -q -F ${2} ${EXTRA_OPTS}"
+            if [ "${5}" = "y" ]; then
+                run_and_log_cmd "${MKFS_EXT4} -q -F /dev/mapper/img ${EXTRA_OPTS}"
+            else
+                run_and_log_cmd "${MKFS_EXT4} -q -F ${2} ${EXTRA_OPTS}"
+            fi
```
Comment (vpodzime) [inline modules/disk.sh:140]: No support with XFS? Looks like handling this inside the `case` is not great.

File: mender-convert-modify [hunk 6]
```diff
@@ -274,7 +274,7 @@ EOF
     run_and_log_cmd "sudo install -m 751 work/grub-editenv work/rootfs/usr/bin/"
 
     run_and_log_cmd "sudo mkdir -p work/boot/EFI/BOOT"
-    run_and_log_cmd "sudo cp work/${efi_name} -P work/boot/EFI/BOOT/${efi_target_name}"
+    run_and_log_cmd "sudo cp grub/${efi_name} -P work/boot/EFI/BOOT/${efi_target_name}"
```
Comment (vpodzime) [inline mender-convert-modify:277]: Looks like this may break cases where `MENDER_GRUB_BINARY_STORAGE_URL` provides some special GRUB.

**Review by vpodzime — COMMENTED** [review]
Looks good to me otherwise. Thank you for the contribution!

**Comment by kacf** [comment]
@nandra: Same here, is this ready for review, or are you still working on it?

**Comment by nandra** [comment]
@kacf added support for static luks UUID but still some tweaks in grub env is missing (not compatible) so still on it + missing state script which set correct UUID when update is one for partition. I hope I can have it ready next week, then will push final changes. Thanks.

**Pull request closed without merging**